controller/src/main/java/io/pravega/controller/store/stream/tables/StreamTruncationRecord.java

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream.tables;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Data class for storing information about stream's truncation point.
+ */
+@AllArgsConstructor(access = AccessLevel.PACKAGE)
+public class StreamTruncationRecord implements Serializable {
+    public static final StreamTruncationRecord EMPTY = new StreamTruncationRecord(ImmutableMap.of(),
+            ImmutableMap.of(), ImmutableSet.of(), ImmutableSet.of());
+
+    /**
+     * Stream cut that is applied as part of this truncation.
+     */
+    private final ImmutableMap<Integer, Long> streamCut;
+
+    /**
+     * If a stream cut spans across multiple epochs then this map captures mapping of segments from the stream cut to
+     * epochs they were found in closest to truncation point.
+     * This data structure is used to find active segments wrt a stream cut.
+     * So for example:
+     * epoch 0: 0, 1
+     * epoch 1: 0, 2, 3
+     * epoch 2: 0, 2, 4, 5
+     * epoch 3: 0, 4, 5, 6, 7
+     *
+     * Following is a valid stream cut {0/offset, 3/offset, 6/offset, 7/offset}
+     * This spans from epoch 1 till epoch 3. Any request for segments at epoch 1 or 2 or 3 will need to have this stream cut
+     * applied on it to find segments that are available for consumption.
+     * Refer to TableHelper.getActiveSegmentsAt
+     */
+    private final ImmutableMap<Integer, Integer> cutEpochMap;
+    /**
+     * All segments that have been deleted for this stream so far.
+     */
+    private final ImmutableSet<Integer> deletedSegments;
```

Comment:
toDelete is for segments deleted for the current stream cut.
deletedSegments keeps all deleted segments across all stream cuts.
This is for optimization so that we do not attempt to delete segments that were already deleted while processing a previous stream cut.

Reply:
Ok, you explained why we need to look at the past and it makes sense, but I'm still not sure whether we are keeping the whole history or not. Say I go through three cuts: S1, S2, S3. I process S1 and delete some segments. When I process S2, I can discard the ones that have been discarded previously with the processing of S1. When I process S3, I don't have to discard the ones that have been discarded with S1 and S2, but do I need to remember the ones that have been deleted with S1? Or, is it the case that we assume that whatever was before the ones deleted in S2 must be gone?

Reply:
Imagine a graph (our typical segment diagram). 
Properties of truncation are:

1. All segments that are predecessor of segments in the stream Cut can be deleted. 
2. A stream cut is valid for truncation iff 
a. it covers the entire key space of 0.0 to 1.0 without any overlaps
b. For each potential key - the new cut is >= previous stream cut. 

So if for a previous cut S1, some segments were deleted (say set D1), they will strictly be predecessors of segments in cut S2. 
So if S1 means deleting D1
S2 means deleting D2
and S3 means deleting D3

Then we only need to delete D2-D1 when we are processing S2 because D1 is already deleted. 
Similarly when we are processing S3, we just need to delete D3 - D2. 

And after completing a truncation we store the overall deleted segments in the truncation record. 
We dont need to keep the history of how we deleted all segments in D3. All we care about is that D3 have been deleted.